cpp/src/arrow/dataset/file_base.h

```diff
@@ -147,11 +153,43 @@ class ARROW_DS_EXPORT FileBasedDataFragment : public DataFragment {
   const FileSource& source() const { return source_; }
   std::shared_ptr<FileFormat> format() const { return format_; }
 
+  std::shared_ptr<ScanOptions> scan_options() const override { return scan_options_; }
```

Comment:
Shouldn't file format be stateless?

Reply:
If so, then we could replace `std::shared_ptr<FileFormat>` with `const FileFormat*` everywhere and let all FileFormat implementations be singletons. Is that what you're suggesting?